<p align="center">
  <img src="https://cdn-icons-png.flaticon.com/512/6295/6295417.png" width="100" />
</p>
<p align="center">
    <h1 align="center">BioCloudLabs BLAST</h1>
</p>
<p align="center">
    <em><code>All-In-One Web to perform BLAST analysis </code></em>
</p>
<p align="center">
	<img src="https://img.shields.io/github/license/BioCloudLabs/blast?style=flat&color=0080ff" alt="license">
	<img src="https://img.shields.io/github/last-commit/BioCloudLabs/blast?style=flat&logo=git&logoColor=white&color=0080ff" alt="last-commit">
	<img src="https://img.shields.io/github/languages/top/BioCloudLabs/blast?style=flat&color=0080ff" alt="repo-top-language">
	<img src="https://img.shields.io/github/languages/count/BioCloudLabs/blast?style=flat&color=0080ff" alt="repo-language-count">
<p>
<p align="center">
		<em>Developed with the software and tools below.</em>
</p>
<p align="center">
	<img src="https://img.shields.io/badge/GNU%20Bash-4EAA25.svg?style=flat&logo=GNU-Bash&logoColor=white" alt="GNU%20Bash">
	<img src="https://img.shields.io/badge/Bootstrap-7952B3.svg?style=flat&logo=Bootstrap&logoColor=white" alt="Bootstrap">
	<img src="https://img.shields.io/badge/Vite-646CFF.svg?style=flat&logo=Vite&logoColor=white" alt="Vite">
	<img src="https://img.shields.io/badge/React-61DAFB.svg?style=flat&logo=React&logoColor=black" alt="React">
	<img src="https://img.shields.io/badge/ESLint-4B32C3.svg?style=flat&logo=ESLint&logoColor=white" alt="ESLint">
	<br>
	<img src="https://img.shields.io/badge/Python-3776AB.svg?style=flat&logo=Python&logoColor=white" alt="Python">
	<img src="https://img.shields.io/badge/Docker-2496ED.svg?style=flat&logo=Docker&logoColor=white" alt="Docker">
	<img src="https://img.shields.io/badge/JSON-000000.svg?style=flat&logo=JSON&logoColor=white" alt="JSON">
	<img src="https://img.shields.io/badge/Flask-000000.svg?style=flat&logo=Flask&logoColor=white" alt="Flask">
</p>
<hr>

## ğŸ”— Quick Links

> - [ğŸ“ Overview](#-overview)
> - [ğŸ“¦ Features](#-features)
> - [ğŸ“‚ Repository Structure](#-repository-structure)
> - [ğŸ§© Modules](#-modules)
> - [ğŸš€ Getting Started](#-getting-started)
>   - [âš™ï¸ Installation](#ï¸-installation)
>   - [ğŸ¤– Running blast](#-running-blast)
>   - [ğŸ§ª Tests](#-tests)
> - [ğŸ›  Project Roadmap](#-project-roadmap)
> - [ğŸ¤ Contributing](#-contributing)
> - [ğŸ“„ License](#-license)
> - [ğŸ‘ Acknowledgments](#-acknowledgments)

---

## ğŸ“ Overview

<code>BLAST is a crucial tool in bioinformatics, enabling researchers to quickly and efficiently identify sequences with high similarity, predict functions of unknown sequences, and explore evolutionary relationships. Its algorithm balances sensitivity and speed, making it suitable for analyzing large datasets.</code>

---

## ğŸ“¦ Features

<code>
â”€ High Speed: Optimized algorithms for rapid results with large databases.
â”€ Adjustable Sensitivity: User-configurable parameters to balance sensitivity and precision.
â”€ Multi-Format Compatibility: Supports various common bioinformatics sequence formats.
â”€ Visual Results: Generates graphical alignments and visualizations.
â”€ Public Database Access: Integrates with numerous public biological sequence databases.
â”€ Filtering Tools: Options to refine and filter searches for specific results.
</code>

---

## ğŸ“‚ Repository Structure

```sh
â””â”€â”€ blast/
    â”œâ”€â”€ Dockerfile
    â”œâ”€â”€ __init__.py
    â”œâ”€â”€ app.py
    â”œâ”€â”€ app.wsgi
    â”œâ”€â”€ blast.py
    â”œâ”€â”€ config.json
    â”œâ”€â”€ container.py
    â”œâ”€â”€ exit.py
    â”œâ”€â”€ httpd.conf
    â”œâ”€â”€ requirements.txt
    â”œâ”€â”€ schema
    â”‚   â”œâ”€â”€ files.py
    â”‚   â””â”€â”€ form.py
    â”œâ”€â”€ start.sh
    â””â”€â”€ static
        â”œâ”€â”€ .eslintrc.cjs
        â”œâ”€â”€ index.html
        â”œâ”€â”€ package.json
        â”œâ”€â”€ public
        â”‚   â””â”€â”€ vite.svg
        â”œâ”€â”€ src
        â”‚   â”œâ”€â”€ App.jsx
        â”‚   â”œâ”€â”€ assets
        â”‚   â”‚   â””â”€â”€ react.svg
        â”‚   â”œâ”€â”€ genomes.js
        â”‚   â””â”€â”€ main.jsx
        â””â”€â”€ vite.config.js
```

---

## ğŸ§© Modules

<details closed><summary>.</summary>

| File                                                                                   | Summary                         |
| ---                                                                                    | ---                             |
| [config.json](https://github.com/BioCloudLabs/blast/blob/master/config.json)           | <code>â–º INSERT-TEXT-HERE</code> |
| [Dockerfile](https://github.com/BioCloudLabs/blast/blob/master/Dockerfile)             | <code>â–º INSERT-TEXT-HERE</code> |
| [start.sh](https://github.com/BioCloudLabs/blast/blob/master/start.sh)                 | <code>â–º INSERT-TEXT-HERE</code> |
| [exit.py](https://github.com/BioCloudLabs/blast/blob/master/exit.py)                   | <code>â–º INSERT-TEXT-HERE</code> |
| [httpd.conf](https://github.com/BioCloudLabs/blast/blob/master/httpd.conf)             | <code>â–º INSERT-TEXT-HERE</code> |
| [blast.py](https://github.com/BioCloudLabs/blast/blob/master/blast.py)                 | <code>â–º INSERT-TEXT-HERE</code> |
| [app.wsgi](https://github.com/BioCloudLabs/blast/blob/master/app.wsgi)                 | <code>â–º INSERT-TEXT-HERE</code> |
| [requirements.txt](https://github.com/BioCloudLabs/blast/blob/master/requirements.txt) | <code>â–º INSERT-TEXT-HERE</code> |
| [app.py](https://github.com/BioCloudLabs/blast/blob/master/app.py)                     | <code>â–º INSERT-TEXT-HERE</code> |
| [container.py](https://github.com/BioCloudLabs/blast/blob/master/container.py)         | <code>â–º INSERT-TEXT-HERE</code> |

</details>

<details closed><summary>schema</summary>

| File                                                                          | Summary                         |
| ---                                                                           | ---                             |
| [files.py](https://github.com/BioCloudLabs/blast/blob/master/schema/files.py) | <code>â–º INSERT-TEXT-HERE</code> |
| [form.py](https://github.com/BioCloudLabs/blast/blob/master/schema/form.py)   | <code>â–º INSERT-TEXT-HERE</code> |

</details>

<details closed><summary>static</summary>

| File                                                                                  | Summary                         |
| ---                                                                                   | ---                             |
| [package.json](https://github.com/BioCloudLabs/blast/blob/master/static/package.json) | <code>â–º INSERT-TEXT-HERE</code> |

</details>

---

## ğŸš€ Getting Started

***Requirements***

Ensure you have the following dependencies installed on your system:

* **Python**: `version x.y.z`

### âš™ï¸ Installation

1. Clone the blast repository:

```sh
git clone https://github.com/BioCloudLabs/blast
```

2. Change to the project directory:

```sh
cd blast
```

3. Install the dependencies:

```sh
pip install -r requirements.txt
```

### ğŸ¤– Running blast

Use the following command to run blast:

```sh
python main.py
```

### ğŸ§ª Tests

To execute tests, run:

```sh
pytest
```

---

## ğŸ›  Project Roadmap

- [X] `â–º INSERT-TASK-1`
- [ ] `â–º INSERT-TASK-2`
- [ ] `â–º ...`

---

## ğŸ¤ Contributing

Contributions are welcome! Here are several ways you can contribute:

- **[Submit Pull Requests](https://github.com/BioCloudLabs/blast/blob/main/CONTRIBUTING.md)**: Review open PRs, and submit your own PRs.
- **[Join the Discussions](https://github.com/BioCloudLabs/blast/discussions)**: Share your insights, provide feedback, or ask questions.
- **[Report Issues](https://github.com/BioCloudLabs/blast/issues)**: Submit bugs found or log feature requests for Blast.

<details closed>
    <summary>Contributing Guidelines</summary>

1. **Fork the Repository**: Start by forking the project repository to your GitHub account.
2. **Clone Locally**: Clone the forked repository to your local machine using a Git client.
   ```sh
   git clone https://github.com/BioCloudLabs/blast
   ```
3. **Create a New Branch**: Always work on a new branch, giving it a descriptive name.
   ```sh
   git checkout -b new-feature-x
   ```
4. **Make Your Changes**: Develop and test your changes locally.
5. **Commit Your Changes**: Commit with a clear message describing your updates.
   ```sh
   git commit -m 'Implemented new feature x.'
   ```
6. **Push to GitHub**: Push the changes to your forked repository.
   ```sh
   git push origin new-feature-x
   ```
7. **Submit a Pull Request**: Create a PR against the original project repository. Clearly describe the changes and their motivations.

Once your PR is reviewed and approved, it will be merged into the main branch.

</details>

---

## ğŸ“„ License

This project is protected under the [SELECT-A-LICENSE](https://choosealicense.com/licenses) License. For more details, refer to the [LICENSE](https://choosealicense.com/licenses/) file.

---

## ğŸ‘ Acknowledgments

- List any resources, contributors, inspiration, etc. here.

[**Return**](#-quick-links)

---
